components/raftstore/src/store/fsm/peer.rs

```diff
@@ -5311,6 +5378,17 @@ where
             replicated_idx
         };
         assert!(compact_idx >= first_idx);
+
+        if let Some(idx) = witness_replicated_idx {
+            // 1. do not compact the log not replicated to witness, except witness is the
+            // most lagging peer
+            // 2. if there is a lagging behind follower, should not compact log, otherwise
```

Comment:
How about delaying witness's compact log? We can include replicated index in compact log request. Witness delete logs only when replicated index > last compact log index.

Reply:
It's complicated
- If there is no new write, replicated index is update but compact log won't be triggered again, then the witness's log won't be compacted in time
- How does the leader know whether the witness has too many logs

Reply:
If hibernate is enabled, there is a negotiate stage of hibernate, replicated index can be set in the message. Or the witness can query the lagging peers whether they have catch up logs and then delete its local logs.